README.md
# Ponse

Module for work with requests and responses

## Install

`npm i ponse --save`

## Api

**isGZIP(request)** check is gzip supported

- http request

**getPathName(request || string)** - get pathname

- http request

**getQuery(request || string)** - get query

- http request

**static(options)** - serve static content

- options:
  - cache
  - root

**send(data, params)** - send data

## License

MIT


ChangeLog
2022.04.23, v7.0.0

feature:
- (ponse) add support of NBSP
- (package) supertape v7.2.3
- (package) eslint v8.14.0
- (package) putout v25.15.0
- (package) eslint-plugin-putout v14.11.0
- (package) madrun v9.0.2
- (ponse) drop support of node < 16


2021.01.25, v6.0.0

fix:
- Initial commit.
- Initial commit.

feature:
- (package) @cloudcmd/stub v3.3.0
- (package) file-type v16.2.0
- (ponse) drop support of node < 14
- (package) madrun v8.6.2
- (package) putout v13.9.1
- (package) eslint-plugin-putout v6.13.1
- (package) eslint v7.18.0
- (package) files-io v4.0.0
- (package) supertape v4.8.1


2020.04.23, v5.1.1

fix:
- (ponse) send: node v14


2020.04.23, v5.1.0

feature:
- (package) rm execon


2020.02.26, v5.0.1

feature:
- (package) serve-once v2.0.0
- (package) try-to-catch v3.0.0


2020.02.21, v5.0.0

feature:
- (package) nyc v15.0.0
- (package) eslint-plugin-node v11.0.0
- (ponse) drop support of node < 10


2019.12.15, v4.0.0

feature:
- (ponse) drop support of directory as first argument in static


2019.12.13, v3.5.1

fix:
- (ponse) typeof -> itype


2019.12.13, v3.5.0

feature:
- (package) got -> serve-once
- (package) nodemon v2.0.2
- (package) eslint-plugin-putout v3.0.0
- (package) got v10.0.4
- (package) eslint-plugin-node v10.0.0
- (package) putout v7.3.1
- (package) madrun v5.0.1
- (package) itype v3.0.1


2019.09.11, v3.4.1

feature:
- (ponse) rm legacy
- (ponse) download: Content-Type: application/octet-stream -> Content-Disposition: attachment


2019.08.23, v3.4.0

feature:
- (ponse) download: Content-Type: application/octet-stream -> Content-Disposition: attachment


2019.08.23, v3.3.1

feature:
- (ponse) forEach -> for-of
- (package) eslint-plugin-node v9.1.0
- (package) eslint v6.2.1
- (package) nyc v14.1.1


2018.11.06, v3.3.0

feature:
- (getQuery) add support of string


2018.11.01, v3.2.0

feature:
- (package) send: name is not required


2018.10.30, v3.1.0

feature:
- (ponse) getPathName: add support of str


2018.10.30, v3.0.2

fix:
- (ponse) rm deprecated module url (https://nodejs.org/dist/latest-v11.x/docs/api/url.html#url_legacy_url_api)

feature:
- (package) eslint-plugin-node v8.0.0
- (ponse) es2018-ify


2018.10.24, v3.0.1

feature:
- (ponse) extendy -> Object.assign


2018.10.24, v3.0.0

feature:
- (package) debug v4.1.0
- (package) redrun v7.0.2
- (package) files-io v3.0.0
- (package) drop support of node < 8
- (package) nyc v13.0.1
- (package) got v9.2.2


2018.07.19, v2.0.3

fix:
- (ponse) static: related path


2018.07.19, v2.0.2

fix:
- (ponse) path traversal
- (ponse) , -> ;

feature:
- (package) add eslint-plugin-node
- (package) eslint-plugin-node v7.0.1
- (package) eslint v5.1.0


2018.07.02, v2.0.1

feature:
- (package) eslint v5.0.1
- (ponse) simplify path to ext.json
- (package) lint


2018.05.02, v2.0.0

feature:
- (package) files-io v2.0.0
- (ponse) drop support of node < 4


2017.08.09, v1.6.1

feature:
- (package) debug v3.0.0


2017.01.18, v1.6.0

feature:
- (ponse) sendError: send stack -> log stack


2017.01.17, v1.5.0

feature:
- (ponse) sendError: send stack if exists


2016.12.28, v1.4.5

feature:
- (package) itype v2.0.0


2016.11.25, v1.4.4

fix:
- (ponse) sendFile: mtime -> mtime.toUTCString

feature:
- (package) ~ -> ^


2016.11.09, v1.4.3

feature:
- (package) debug v2.3.0


2016.06.07, v1.4.2

fix:
- (ext) proper mime type for eot


2015.06.10, v1.4.1

feature:
- (ponse) rm console.log


2015.06.10, v1.4.0

feature:
- (ponse) send, sendFile, setHeader: add mime


2015.06.03, v1.3.6

feature:
- (package) execon v1.2.0


2015.05.12, v1.3.5

feature:
- (package) debug v2.2.0


2015.02.02, v1.3.4

feature:
- (ponse) checkup -> assert


2015.01.16, v1.3.3

feature:
- (package) extend -> extendy


2015.01.15, v1.3.2

feature:
- (package) execon v1.1.0


2014.12.22, v1.3.1

feature:
- (package) rm util-io; add execon, checkup, extend, itype


2014.12.08, v1.3.0

feature:
- (ponse) send: rm notLog


2014.11.18, v1.2.8

fix:
- (ponse) ./files -> files-io


2014.11.18, v1.2.7

fix:
- (ponse) ../util -> util-io


2014.11.17, v1.2.6

feature:
- (ponse) rm unnecessary conditions


2014.11.03, v1.2.5

fix:
- (ponse) DIR_JSON


2014.11.03, v1.2.4

feature:
- (ponse) pipe-io -> files-io


2014.10.31, v1.2.3

feature:
- (package) util-io v1.5.0
- (package) pipe-io v1.0.0



test/ponse.js
'use strict';

const {test, stub} = require('supertape');

const ponse = require('..');

const {
    getPathName,
    getQuery,
    send,
} = ponse;

const {request} = require('serve-once')(ponse.static);

test('ponse: path traversal: status', async (t) => {
    const {status} = await request.get('/../../../../../../etc/passwd', {
        options: {
            root: __dirname,
        },
    });
    
    t.equal(status, 404);
    t.end();
});

test('ponse: path traversal: message', async (t) => {
    const options = {
        root: __dirname,
    };
    const request = {
        url: '/../../../../../../../../etc/passwd',
    };
    
    const response = {
        send: stub(),
        setHeader: stub(),
        end: stub(),
    };
    
    await ponse._getStatic(options, request, response);
    const expect = `Path /etc/passwd beyond root ${__dirname}!`;
    
    t.calledWith(response.end, [expect]);
    t.end();
});

test('ponse: path traversal: status: ok', async (t) => {
    const {status} = await request.get('/ponse.js', {
        options: {
            root: __dirname,
        },
    });
    
    t.equal(status, 200);
    t.end();
});

test('ponse: path traversal: status: ok: changed root', async (t) => {
    const {status} = await request.get('/ponse.js', {
        options: {
            root: `${__dirname}/../lib`,
        },
    });
    
    t.equal(status, 200);
    t.end();
});

test('ponse: getPathName: res', (t) => {
    const url = '/hello?world=1';
    const name = getPathName({
        url,
    });
    
    t.equal(name, '/hello');
    t.end();
});

test('ponse: getPathName: nbsp', (t) => {
    const url = '/hello&nbsp;world?world=1';
    const name = getPathName({
        url,
    });
    
    t.equal(name, '/hello\xa0world');
    t.end();
});

test('ponse: getPathName: string', (t) => {
    const url = '/hello?world=1';
    const name = getPathName({
        url,
    });
    
    t.equal(name, '/hello');
    t.end();
});

test('ponse: getQuery', (t) => {
    const req = {
        url: 'hi?world',
    };
    
    const query = getQuery(req);
    
    t.equal(query, 'world');
    t.end();
});

test('ponse: getQuery: string', (t) => {
    const url = 'hi?world';
    const query = getQuery(url);
    
    t.equal(query, 'world');
    t.end();
});

test('ponse: send', (t) => {
    const request = {};
    const response = {
        writableFinished: true,
        end: stub(),
        setHeader: stub(),
    };
    
    send('hello', {
        response,
        request,
    });
    
    t.notOk(response.end.called, 'should not call response.end');
    t.end();
});

test('ponse: get: content-type', async (t) => {
    const filesToMimes = {
        '320px-Floppy_disk_2009_G1': 'image/jpeg',
        '320px-Floppy_disk_2009_G1.jpg': 'image/jpeg',
        '294px-Railroad1860.png': 'image/png',
        '294px-Railroad1860': 'image/png',
        'hello.txt': 'text/plain; charset=UTF-8',
    };
    
    let file;
    
    for (file in filesToMimes) {
        const {headers} = await request.get('/' + file, {
            options: {
                root: __dirname + '/../test/fixtures/mimetype/',
            },
        });
        
        const reportedMime = headers.get('content-type');
        const expect = filesToMimes[file];
        
        t.equal(reportedMime, expect);
    }
    
    t.end();
}, {
    checkAssertionsCount: false,
});


test/connect.js
'use strict';

const http = require('http');

const got = require('got');
const {promisify} = require('util');
const ponse = require('..');

const getURL = (path, port) => `http://127.0.0.1:${port}/${path}`;

const serve = promisify((dir, fn) => {
    const server = http.createServer(ponse.static(dir));
    const done = () => server.close();
    
    server.listen(() => {
        const {port} = server.address();
        
        fn(null, {
            port,
            done,
        });
    });
});

module.exports.serve = serve;

module.exports.get = async (path, root) => {
    const {port, done} = await serve(root);
    const url = getURL(path, port);
    
    const res = await got(url, {
        throwHttpErrors: false,
    });
    
    done();
    
    return res;
};



lib/ponse.js
'use strict';

const zlib = require('zlib');
const path = require('path');
const assert = require('assert');
const {promisify} = require('util');
const {unescape} = require('querystring');
const {lstat} = require('fs/promises');

const type = require('itype');
const filetype = require('file-type');
const tryToCatch = require('try-to-catch');
const files = require('files-io');

const debug = require('debug');
const logData = debug('ponse:data');
const logError = debug('ponse:error');

const ext = require('../json/ext');

const gzip = promisify(zlib.gzip);

const OK = 200;
const RANGE = 206;
const MOVED_PERMANENTLY = 301;
const FILE_NOT_FOUND = 404;

const KILOBYTE = 1024;

const {assign} = Object;

exports.redirect = redirect;

exports.send = send;
exports.sendError = sendError;
exports.sendFile = sendFile;

exports.isGZIP = isGZIP;

exports.getPathName = getPathName;
exports.getQuery = getQuery;

exports.setHeader = setHeader;

exports.static = (options) => getStatic.bind(null, options);

/* Функция высылает ответ серверу
 * @param data
 * @param params
 * @param notLog
 */
async function send(data, params) {
    const p = params;
    const {
        name = '',
        cache,
        query,
        mime,
        response,
    } = params;
    
    checkParams(params);
    
    if (response.writableFinished)
        return;
    
    const isGzip = p.gzip && isGZIP(p.request) && data.length > KILOBYTE;
    const head = generateHeaders({
        name,
        cache,
        query,
        mime,
        gzip: isGzip,
    });
    
    fillHeader(head, response);
    
    logData(data);
    
    response.statusCode = p.status || OK;
    
    /* если браузер поддерживает gzip-сжатие - сжимаем данные*/
    if (!isGzip)
        return response.end(data);
    
    const [error, zipped] = await tryToCatch(gzip, data);
    
    if (!error)
        return response.send(zipped);
    
    response.statusCode = FILE_NOT_FOUND;
    response.end(error.message);
}

/**
 * Функция создаёт заголовки файлов
 * в зависимости от расширения файла
 * перед отправкой их клиенту
 * @param pParams
 *  name - имя файла
 * gzip - данные сжаты gzip'ом
 * query
 * https://developers.google.com/speed/docs/best-practices/caching?hl=ru#LeverageProxyCaching
 */
function generateHeaders(params) {
    let maxAge = 31_337 * 21;
    
    const p = params;
    const extension = path.extname(p.name);
    const type = p.mime || ext[extension] || 'text/plain';
    let encoding = '';
    
    if (!/img|image|audio|video/.test(type))
        encoding = '; charset=UTF-8';
    
    const header = {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': type + encoding,
        'Vary': 'Accept-Encoding',
        'Accept-Ranges': 'bytes',
    };
    
    if (/download/.test(p.query)) {
        header['Content-Disposition'] = 'attachment';
    }
    
    if (p.time)
        assign(header, {
            'Last-Modified': p.time,
        });
    
    if (p.range)
        assign(header, {
            'Content-Range': 'bytes ' + p.range.start +
                                '-' + p.range.end +
                                '/' + p.range.sizeTotal,
            
            'Content-Length': p.range.size,
        });
    else if (p.length)
        assign(header, {
            'Content-Length': p.length,
        });
    
    const cmp = extension === '.appcache';
    
    if (!p.cache || cmp)
        maxAge = 0;
    
    header['Cache-Control'] = 'max-age=' + maxAge;
    
    if (p.gzip)
        header['Content-Encoding'] = 'gzip';
    
    return header;
}

function setHeader(params) {
    const p = params;
    
    checkParams(params);
    
    const gzip = isGZIP(p.request) && p.gzip;
    const header = generateHeaders({
        name: p.name,
        time: p.time,
        range: p.range,
        length: p.length,
        cache: p.cache,
        mime: p.mime,
        gzip,
        query: getQuery(p.request),
    });
    
    fillHeader(header, p.response);
    p.response.statusCode = p.status || OK;
}

function fillHeader(header, response) {
    const isObject = type.object(header);
    const isSent = response.headersSent;
    
    if (!isSent && isObject)
        for (const name of Object.keys(header)) {
            response.setHeader(name, header[name]);
        }
}

/**
 * send file to client thru pipe
 * and gzip it if client support
 *
 */
async function sendFile(params) {
    checkParams(params);
    
    const p = params;
    const [error, stat] = await tryToCatch(lstat, p.name);
    
    if (error)
        return await sendError(error, params);
    
    const isGzip = isGZIP(p.request) && p.gzip;
    const time = stat.mtime.toUTCString();
    const length = stat.size;
    const range = getRange(p.request, length);
    
    if (range)
        assign(p, {
            range,
            status: RANGE,
        });
    
    assign(p, {
        time,
    });
    
    if (!isGzip)
        p.length = length;
    
    const [ftError, ftype] = await tryToCatch(filetype.fromFile, p.name);
    
    if (!ftError && ftype)
        p.mime = ftype.mime;
    
    setHeader(params);
    
    const options = {
        gzip: isGzip,
        range,
    };
    
    const [pipeError] = await tryToCatch(files.pipe, p.name, p.response, options);
    
    if (pipeError)
        await sendError(pipeError, params);
}

/**
 * send error response
 */
async function sendError(error, params) {
    checkParams(params);
    
    params.status = FILE_NOT_FOUND;
    
    const data = error.message || String(error);
    
    logError(error.stack);
    
    await send(data, params);
}

function checkParams(params) {
    const ERROR = 'could not be empty!';
    const p = params;
    
    assert(params, 'params ' + ERROR);
    assert(p.request, 'p.request ' + ERROR);
    assert(p.response, 'p.response ' + ERROR);
}

function getQuery(req) {
    assert(req, 'req could not be empty!');
    
    const {
        url = req,
    } = req;
    return url.replace(/^.*\?/, '');
}

function cutQuery(url) {
    return url.replace(/\?.*/, '');
}

function getPathName(req) {
    assert(req, 'req could not be empty!');
    const url = req.url.replace('&nbsp;', '\xa0');
    
    const pathname = cutQuery(url);
    // supporting of Russian language in directory names
    return unescape(pathname);
}

function getRange(req, sizeTotal) {
    let range;
    let start;
    let end;
    let size;
    let parts;
    const rangeStr = req.headers.range;
    
    if (rangeStr) {
        parts = rangeStr.replace('bytes=', '').split('-');
        [start] = parts;
        end = parts[1] || sizeTotal - 1;
        size = end - start + 1;
        
        range = {
            start: start - 0,
            end: end - 0,
            size,
            sizeTotal,
        };
    }
    
    return range;
}

function isGZIP(req) {
    const enc = req.headers['accept-encoding'] || '';
    const is = /\bgzip\b/.test(enc);
    
    return is;
}

/**
 * redirect to another URL
 */
function redirect(url, response) {
    const header = {
        Location: url,
    };
    
    assert(url, 'url could not be empty!');
    assert(response, 'response could not be empty!');
    
    fillHeader(header, response);
    response.statusCode = MOVED_PERMANENTLY;
    response.end();
}

module.exports._getStatic = getStatic;
async function getStatic(options, request, response) {
    const o = options || {};
    const {root} = options;
    const dir = path.normalize(root);
    const pathName = getPathName(request);
    const name = path.join(dir, pathName);
    
    if (name.indexOf(dir)) {
        return await sendError(Error(`Path ${name} beyond root ${dir}!`), {
            request,
            response,
            name,
        });
    }
    
    let cache;
    
    if (type.function(o.cache))
        cache = o.cache();
    else if (o.cache !== undefined)
        cache = o.cache;
    else
        cache = true;
    
    const gzip = true;
    
    await sendFile({
        name,
        cache,
        gzip,
        request,
        response,
    });
}

